ckb-lock-common/src/generate_sighash_all.rs

```diff
@@ -1,118 +1,79 @@
-    }
-    // Clear lock field to zero, then digest the first witness
-    // lock_bytes_seg.ptr actually points to the memory in temp buffer.
-    temp[20..20 + lock_length].fill(0);
-
-    // Load tx hash.
-    let tx_hash = load_tx_hash()?;
-
-    // Prepare sign message.
-    let mut blake2b_ctx = new_blake2b();
-    blake2b_ctx.update(&tx_hash);
-    blake2b_ctx.update(&(witness_len as u64).to_le_bytes());
-    blake2b_ctx.update(&temp[..read_len]);
-
-    // Remaining of first witness.
-    if read_len < witness_len {
-        load_and_hash_witness(&mut blake2b_ctx, read_len, 0, Source::GroupInput, false)?;
-    }
+    // Digest first witness in the script group.
+    let mut chunks = ChunksLoader::new(load_witness, CHUNK_SIZE, 0, Source::GroupInput).into_iter();
+    let mut ctx = new_blake2b();
+    if let Some((total_len, mut chunk_data)) = chunks.next() {
+        if total_len < 20 {
+            return Err(Error::Encoding);
+        }
+        let lock_length = u32::from_le_bytes(chunk_data[16..20].try_into().unwrap()) as usize;
+        if total_len < lock_length + 20 {
+            return Err(Error::Encoding);
+        }
+        let mut zero_remain = lock_length + 20;
+        if zero_remain > chunk_data.len() {
+            chunk_data[20..].fill(0);
+            zero_remain -= CHUNK_SIZE;
+        } else {
+            chunk_data[20..zero_remain].fill(0);
+            zero_remain = 0;
+        }
 
-    // Digest same group witnesses.
-    let mut i = 1;
-    loop {
-        let sysret = load_and_hash_witness(&mut blake2b_ctx, 0, i, Source::GroupInput, true);
-        match sysret {
-            Err(SysError::IndexOutOfBound) => break,
-            Err(x) => return Err(x.into()),
-            Ok(_) => i += 1,
+        let tx_hash = load_tx_hash()?;
+        ctx.update(&tx_hash);
+        ctx.update(&(total_len as u64).to_le_bytes());
+        ctx.update(&chunk_data);
+        for (_total_len, mut chunk_data) in chunks {
+            if zero_remain > 0 {
+                if zero_remain > chunk_data.len() {
+                    chunk_data[..].fill(0);
+                    zero_remain -= CHUNK_SIZE;
+                } else {
+                    chunk_data[..zero_remain].fill(0);
+                    zero_remain = 0;
+                }
+            }
+            ctx.update(&chunk_data);
         }
+    } else {
+        return Err(Error::Encoding);
     }
 
+    // Digest other witnesses in the script group.
+    load_and_hash_witness(&mut ctx, 1, Source::GroupInput);
     // Digest witnesses that not covered by inputs.
-    let mut i = calculate_inputs_len()?;
-
-    loop {
-        let sysret = load_and_hash_witness(&mut blake2b_ctx, 0, i, Source::Input, true);
-        match sysret {
-            Err(SysError::IndexOutOfBound) => break,
-            Err(x) => return Err(x.into()),
-            Ok(_) => i += 1,
-        }
-    }
+    let i = calculate_inputs_len()?;
+    load_and_hash_witness(&mut ctx, i, Source::Input);
```

Comment:
Need to hash all remaining witnesses started from index `i`, not only one.

Reply:
there is a loop in fn `load_and_hash_witness` to hash all witnesses starting from the index i.